### DISCUSSION TITLE
SSL configuration API could be improved.

### DISCUSSION THREAD

**Discussion by lovelydinosaur** [encode/httpx#2521] [General]
Our SSL configuration currently uses the following flags:

* `verify` - Either `True`, `False`, or a `<ca_bundle_path>`, or an `SSLContext`.
* `cert` - Either a `<cert_file>` or a `(<cert_file>, <key_file>)` or a `(<cert_file>, <key_file>, <password>)`.
* `trust_env` - Either `True` or `False`. Enables `SSLKEYLOGFILE`, `SSL_CERT_FILE`, `SSL_CERT_DIR`.
* `http2` - Sets [the `alpn` protocols](https://github.com/encode/httpx/blob/10a3b68a71d6cb16b170dec4efe2ec1ae197d65f/httpx/_config.py#L165-L166) to allow HTTP/2 negotiation.

We also have the helper `httpx.create_ssl_context()` function available.

We have this API because we've followed in the footsteps of `requests`, but I find the various overloadings a bit confusing, and I'm not wild about having a whole bunch of different configurations at the top-level, when we could instead define them in a way that more tightly binds them to the scope where they're actually used.

I've wondered if we'd provide a better UX if we instead had...

```python
response = httpx.get(..., ssl=<ssl_context>)
```

And...

```python
ssl_context = httpx.SSLContext(verify=..., cert=..., trust_env=..., http2=...)
response = httpx.get(..., ssl=ssl_context)
```

This also makes alternate options more obvious...

```python
ssl_context = truststore.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
response = httpx.get(..., ssl=ssl_context)
```

Which we *can* currently do that just fine with `verify=...` but it isn't an obvious API to me, given the overloaded meanings.

There's some further considerations we'd need to make here:

* Is changing the API worth it?
* `trust_env` is *also* used elsewhere in the client (proxy config and `netrc` configuration I think).
* `http2` config is *also* passed through to the default transport.

I'm not sure this is worth us moving on, but maybe?...

**Comment by Kludex** [comment]
I'm going to provide some "analogous" discussion reference from `uvicorn` - just because we are also trying to improve the SSL context related settings over there, but the solution there is probably not a good one for here: https://github.com/encode/uvicorn/pull/1815#issuecomment-1364716832.
